Firestore/core/src/firebase/firestore/remote/serializer.cc

```diff
@@ -166,10 +173,12 @@ Writer Writer::Wrap(std::vector<uint8_t>* out_bytes) {
 // together (probably within their own file rather than here).
 
 void Writer::WriteTag(pb_wire_type_t wiretype, uint32_t field_number) {
-  bool status = pb_encode_tag(&stream_, wiretype, field_number);
-  if (!status) {
-    // TODO(rsgowman): figure out error handling
-    abort();
+  if (!status_.ok()) return;
+
+  if (!pb_encode_tag(&stream_, wiretype, field_number)) {
+    const char* errmsg = PB_GET_ERROR(&stream_);
+    FIREBASE_DEV_ASSERT_MESSAGE(false, errmsg);
```

Comment:
I don't understand why this is a desirable setup.

Either
  * there should never be an error and we're not prepared to handle it in which case we should assert
  * errors are possible and we should wrap them in a Status and be ready to handle them and should have tests that exercise them.

This arrangement seems like the worst possible setup whereby this could generate errors and then follow a code path that never happens for developers of the software.

Reply:
It's not quite either of those. There should never be an error, but we are prepared to handle it.

If we get here, it indicates either a serious bug in our code, or a memset() or equivalent gone wrong. My expectation is that for any FirestoreErrorCode::Internal related error, that the calling code effectively aborts.

So from that perspective, it might be reasonable to never return a `Internal` error and instead always assert false. However, it's possible that the end dev's code might want to do some additional logging before crashing or maybe display a message to the user; this gives them the opportunity to do so. But for us, I'd rather crash immediately to aid in debugging. (Especially since status objects don't track where they were created.)

If you don't buy that, then we could:

1) Never use FirestoreErrorCode::Internal, and always assert false. Pros: This should indicate exactly where the crash is and the user code can't do anything reasonable except crash in this case anyway. Cons: User code doesn't get the chance to add additional context before crashing.
2) Never assert false. Pros: same code path. Cons: More difficult to find the location of the problem since we won't crash until later.
3) As 2, but create an error message with more context. `errmsg = "Writer::WriteTag failed while encoding a tag via nanopb: " + errmsg;` or similar. (Alternatively, StatusBuilder seems to have the ability to add a log with a stack trace.)

Reply:
Implemented 1 here: https://github.com/firebase/firebase-ios-sdk/pull/1062

> Accordingly I think this means we don't really have any use for FIREBASE_DEV_ASSERT.

FYI: In master:
```
$ grep -r FIREBASE_DEV_ASSERT . | wc -l
35
```

(though 6 of them will be removed due to the above PR.)